File: PiDay/Core/Domain/SearchFormatPreference.swift
import Foundation

// WHY: This enum maps a user-facing preference ("I use DD/MM") to the concrete
// DateFormatOptions the repository should search. Domain-only — no SwiftUI import needed.
enum SearchFormatPreference: String, CaseIterable, Identifiable {
    case international
    case american
    case iso8601
    case all

    var id: String { rawValue }

    var label: String {
        switch self {
        case .international:
            return "DD/MM"
        case .american:
            return "MM/DD"
        case .iso8601:
            return "ISO"
        case .all:
            return "All"
        }
    }

    var title: String {
        switch self {
        case .international:
            return "International"
        case .american:
            return "American"
        case .iso8601:
            return "ISO 8601"
        case .all:
            return "Indexed Formats"
        }
    }

    var formats: [DateFormatOption] {
        switch self {
        case .international:
            // WHY both: DDMMYYYY and D/M/YYYY are the same date order — one with
            // and one without leading zeros. Both are correct international forms,
            // and a date that isn't found in DDMMYYYY may still appear in D/M/YYYY.
            return [.ddmmyyyy, .dmyNoLeadingZeros]
        case .american:
            return [.mmddyyyy]
        case .iso8601:
            return [.yyyymmdd]
        case .all:
            // WHY all five: "Indexed Formats" means every format in the bundled index.
            // Previously only three were searched, meaning users could be told "not found"
            // when their date existed in YYMMDD or D/M/YYYY form.
            return [.ddmmyyyy, .dmyNoLeadingZeros, .mmddyyyy, .yyyymmdd, .yymmdd]
        }
    }

    // The single "hero" format shown prominently when no match is found.
    var heroFormat: DateFormatOption {
        switch self {
        case .international, .all:
            return .ddmmyyyy
        case .american:
            return .mmddyyyy
        case .iso8601:
            return .yyyymmdd
        }
    }

    // Short label used in "No exact X hit" messages.
    var summary: String {
        switch self {
        case .international:
            return "DDMMYYYY or D/M/YYYY"
        case .american:
            return "MMDDYYYY"
        case .iso8601:
            return "YYYYMMDD"
        case .all:
            return "any indexed format"
        }
    }
}

File: PiDayAndroid/app/src/main/java/academy/glasscode/piday/core/domain/SearchFormatPreference.kt
package academy.glasscode.piday.core.domain

enum class SearchFormatPreference {
    INTERNATIONAL, AMERICAN, ISO8601, ALL;

    val label: String get() = when (this) {
        INTERNATIONAL -> "DD/MM"
        AMERICAN      -> "MM/DD"
        ISO8601       -> "ISO"
        ALL           -> "All"
    }

    val title: String get() = when (this) {
        INTERNATIONAL -> "International"
        AMERICAN      -> "American"
        ISO8601       -> "ISO 8601"
        ALL           -> "Indexed Formats"
    }

    val formats: List<DateFormatOption> get() = when (this) {
        INTERNATIONAL -> listOf(DateFormatOption.DDMMYYYY, DateFormatOption.DMY_NO_LEADING_ZEROS)
        AMERICAN      -> listOf(DateFormatOption.MMDDYYYY)
        ISO8601       -> listOf(DateFormatOption.YYYYMMDD)
        ALL           -> listOf(
            DateFormatOption.DDMMYYYY, DateFormatOption.DMY_NO_LEADING_ZEROS,
            DateFormatOption.MMDDYYYY, DateFormatOption.YYYYMMDD, DateFormatOption.YYMMDD
        )
    }

    val heroFormat: DateFormatOption get() = when (this) {
        INTERNATIONAL, ALL -> DateFormatOption.DDMMYYYY
        AMERICAN           -> DateFormatOption.MMDDYYYY
        ISO8601            -> DateFormatOption.YYYYMMDD
    }

    val summary: String get() = when (this) {
        INTERNATIONAL -> "DDMMYYYY or D/M/YYYY"
        AMERICAN      -> "MMDDYYYY"
        ISO8601       -> "YYYYMMDD"
        ALL           -> "any indexed format"
    }
}
